coalib/core/Bear.py

```diff
@@ -0,0 +1,502 @@
+
+        >>> from dependency_management.requirements.PipRequirement import (
+        ...     PipRequirement)
+        >>> class SomeBear(Bear):
+        ...     REQUIREMENTS = {PipRequirement('pip')}
+
+        >>> SomeBear.check_prerequisites()
+        True
+
+        >>> class SomeOtherBear(Bear):
+        ...     REQUIREMENTS = {PipRequirement('really_bad_package')}
+
+        >>> SomeOtherBear.check_prerequisites()
+        'Following requirements are not installed: really_bad_package (...)'
+
+        :return: True if prerequisites are satisfied, else False or a string
+                 that serves a more detailed description of what's missing.
+        """
+        not_installed_requirements = [requirement
+                                      for requirement in cls.REQUIREMENTS
+                                      if not requirement.is_installed()]
+
+        if not_installed_requirements:
+            return 'Following requirements are not installed: ' + ', '.join(
+                '{} (installable via `{}`)'.format(
+                    requirement.package,
+                    ' '.join(requirement.install_command()))
+                for requirement in not_installed_requirements)
+        else:
+            return True
+
+    def get_config_dir(self):
+        """
+        Gives the directory where the configuration file resides.
+
+        :return:
+            Directory of the config file.
+        """
+        return get_config_directory(self.section)
+
+    @classmethod
+    def download_cached_file(cls, url, filename):
+        """
+        Downloads the file if needed and caches it for the next time. If a
+        download happens, the user will be informed.
+
+        Take a sane simple bear:
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+
+        We can now carelessly query for a neat file that doesn't exist yet:
+
+        >>> from os import remove
+        >>> if exists(join(bear.data_dir, 'a_file')):
+        ...     remove(join(bear.data_dir, 'a_file'))
+        >>> file = bear.download_cached_file('https://github.com/', 'a_file')
+
+        If we download it again, it'll be much faster as no download occurs:
+
+        >>> newfile = bear.download_cached_file(
+        ...     'https://github.com/', 'a_file')
+        >>> newfile == file
+        True
+
+        :param url:
+            The URL to download the file from.
+        :param filename:
+            The filename it should get, e.g. "test.txt".
+        :return:
+            A full path to the file ready for you to use!
+        """
+        filename = join(cls.data_dir, filename)
+        if exists(filename):
+            return filename
+
+        logging.info('{}: Downloading {} into {!r}.'
+                     .format(cls.name, url, filename))
+
+        response = requests.get(url, stream=True, timeout=20)
+        with open(filename, 'wb') as file:
```

Comment:
This was implemented for the old bear on master and adapted in https://github.com/coala/coala/pull/3711 :+1: This function now properly throws an exception.